---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# R execution for Seurat
Below is the standard execution for 10X single cells libraries processed through cellrangerv3. 
## Libraries
```{r}
library(Seurat);library(dplyr);library(hdf5r);library(cowplot);
```
```{r}
sessionInfo();
```
## generating the sample list
```{r}
opt <- read.csv(file='metadata.csv')
adt_feats <- read.csv(file='features.csv')
samplelist <- list();
dir.create('QC');
dir.create('clustering');
dir.create('cluster_markers');
dir.create('Gene_figures');
dir.create('Feature_figures');

```

```{r}
cellcounts <- data.frame(matrix(ncol=2,nrow=0)); colnames(cellcounts) <- c('type','count')
for (i in 1:nrow(opt)){
   row <- opt[i,]
   filename <- row$samples
   samp <- row$name
   samp<-paste(samp)
   samptype <- row$type
   tags <- row$tags
   
   directory <- row$directory
   
    ###Import raw data
    data_in <- Read10X_h5(filename=paste(directory,filename,sep=''), use.names=TRUE, unique.features=TRUE);
    #adt_data <- data_in$`Antibody Capture`
    #data_in <- data_in$`Gene Expression`
    
    #adt_data <- Read10X_h5(filename=paste(directory,tags,sep=''),use.names=TRUE, unique.features=TRUE)
    
    ###Create Seurat object
    s <- CreateSeuratObject(counts=data_in[['Gene Expression']],project=,min.cells=3, min.features=200, assay='RNA');
    
    ### Add metadata tags
    s@meta.data$samp <- paste(samp);
    
    s[['ADT']] <- CreateAssayObject(data_in[['Antibody Capture']][, colnames(x = s)])
    
    DefaultAssay(s) <- 'RNA'
    
    ### Mitochondrial
    Do note- if this is human based you need "^MT-", for mouse its "^mt-"
    s[['percent.mito']] <- PercentageFeatureSet(s, pattern='^MT-');
    
    ### Ribosomal
    s[['percent.ribo']] <- PercentageFeatureSet(s,pattern='^RP[SL][[:digit:]]');
    
    ## Visualize pre-filter
    ### Scatter plots
    jpeg(paste('./QC/',samp,'_pre_filt_geneplot_qc.jpg',sep=''))
    in_plot <- FeatureScatter(s, feature1='nCount_RNA',feature2='nFeature_RNA');
    par(mfrow=c(2,2))
    print(in_plot)
    dev.off()
    
    ### Violin/Dot plots
    jpeg(paste('./QC/',samp,'_pre_filt_vln_qc.jpg',sep=''))
    in_plot <- VlnPlot(s, features=c('nFeature_RNA','nCount_RNA','percent.mito','percent.ribo'),ncol=4);
    print(in_plot)
    dev.off()

     ###Cell count pre-filter
    cellcounts[nrow(cellcounts)+1,] <- list(sample=paste(samp, '_prefilt',sep=''), count=length(WhichCells(s)))
    
    ## Filter
    s <- subset (s, subset= nFeature_RNA < 7000 & nFeature_RNA >1000 & percent.mito < 10 & percent.ribo < 45);
    
    ### Violin/Dot plots
    jpeg(paste('./QC/',samp,'_post_filt_vln_qc.jpg',sep=''))
    out_plot <- VlnPlot(s, features=c('nFeature_RNA','nCount_RNA','percent.mito','percent.ribo'),ncol=4);
    print(out_plot)
    dev.off()
    ### Scatter plots
    jpeg(paste('./QC/',samp,'_post_filt_geneplot_qc.jpg',sep=''))
    out_plot <- FeatureScatter(s, feature1='nCount_RNA',feature2='nFeature_RNA');
    print(out_plot)
    dev.off()
    ###Cell count post-filter
    cellcounts[nrow(cellcounts)+1,] <- list(sample=paste(samp, '_postfilt',sep=''), count=length(WhichCells(s)))
    
    ## Transform and regress the data
    s <- SCTransform(s, assay='RNA',vars.to.regress=c('percent.mito','percent.ribo'));
    
    s <- NormalizeData(s, assay='ADT', normalization.method='CLR')
    s <- ScaleData(s, assay='ADT')

    ## Rename temp files
    assign(paste(samp, '.data', sep=''), data_in);
    assign(paste(samp),s);
    
    samplelist <- c(samplelist, eval(parse(text=samp)))
   
   
}
write.csv(cellcounts, './QC/cellcounts_prepost_filter.csv')
```
## Identify the anchor points  and integrate the samples

### Stupid little thing...
The default settings arent fully compliant with this, so you need to adjust the memory availability. I upgrade it to 10Gb, but you can adjust as needed
```{r}
options(future.globals.maxSize=10000*1024^2);
```

### Back to the real stuff... Anchors and integration
I use the standard options in most cases. This pulls 30 dimensions of the dataset. It can be adjusted, though at a certain point you've captured all of the biological heterogeneity and start getting into noise.
```{r}
combined.features <- SelectIntegrationFeatures(object.list=samplelist, nfeatures=3000);
combined.list <- PrepSCTIntegration(object.list=samplelist, anchor.features=combined.features, verbose=FALSE);
combined.anchors <- FindIntegrationAnchors(object.list=combined.list, normalization.method='SCT', anchor.features=combined.features, verbose=FALSE);
combined <- IntegrateData(anchorset=combined.anchors, normalization.method='SCT');
```

## Run the PCA, Find Clusters, and plot it out
Again, using a standard of 30 dimensions. The other facet is the resolution. 0.5 is a good starting point, allows for pulling apart major clusters. If you can visually see significant clusters, increase the resolution to try and isolate them in an unsupervised approach. 
```{r}
combined <- RunPCA(combined, verbose=FALSE, npcs=30);
combined <- RunUMAP(combined, reduction='pca', dims=1:30); 
combined <- FindNeighbors(combined, reduction='pca', dims=1:30); 
combined <- FindClusters(combined, resolution=0.5); 
p1 <- DimPlot(combined, reduction='umap', group.by='samp');
p2 <- DimPlot(combined, reduction='umap', label=TRUE); 
jpeg('./clustering/umap.jpg',height=800, width=1500)
plot_grid(p1,p2); 
dev.off()
jpeg('./clustering/umap_samples.jpg',,height=800, width=1500)
DimPlot(combined, reduction='umap',split.by='samp');
dev.off()
```


```{r}
write.csv(table(Idents(combined),combined$samp), file='./clustering/counts_per_cluster.csv')

```

```{r}
clustnum <- nrow(table(Idents(combined),combined$samp))
for (i in 0:(clustnum-1)){
  clust.markers <- FindConservedMarkers(combined, ident.1=i, grouping.var='samp', print.bar=FALSE)
  write.csv(clust.markers, file=paste('./cluster_markers/clust',i,'.csv',sep=''))
  assign(paste('clust', i,'.markers', sep=''), clust.markers)
  rm(clust.markers)
  }

```

```{r}
DefaultAssay(combined) <- 'SCT'
for (i in 0:(clustnum-1)){
  genelist <- c(paste(rownames(head(get(paste('clust',i,'.markers',sep=''))))))
  for (gene in genelist){
    run <- try({
    jpeg(paste('./Gene_figures/',gene,'_vln_clusts.jpg',sep=''))
    plot <- VlnPlot(combined, features=c(gene))
    print(plot)
    dev.off()
    
    jpeg(paste('./Gene_figures/',gene,'_vln_samp.jpg',sep=''))
    plot <- VlnPlot(combined, features=c(gene), group.by='samp')
    print(plot)
    dev.off()
    
    jpeg(paste('./Gene_figures/',gene,'_feature_plot.jpg',sep=''))
    plot <- FeaturePlot(combined, features=c(gene), min.cutoff='q9')
    print(plot)
    dev.off()

    jpeg(paste('./Gene_figures/',gene,'_ridge_clusts.jpg',sep=''))
    plot <- RidgePlot(combined, features=c(gene))
    print(plot)
    dev.off()

    jpeg(paste('./Gene_figures/',gene,'_ridge_samp.jpg',sep=''))
    plot <- RidgePlot(combined, features=c(gene),group.by='samp')
    print(plot)
    dev.off()}
    )
    if(inherits(run, 'try-error')) { next }
  }
}
```
```{r}
DefaultAssay(combined) <- 'ADT'
for (i in 1:nrow(adt_feats)){
   row <- adt_feats$id
   for (feat in row){
   run <- try({
    jpeg(paste('./Feature_figures/',feat,'_vln_clusts.jpg',sep=''))
    plot <- VlnPlot(combined, features=c(feat))
    print(plot)
    dev.off()
    
    jpeg(paste('./Feature_figures/',feat,'_vln_samp.jpg',sep=''))
    plot <- VlnPlot(combined, features=c(feat), group.by='samp')
    print(plot)
    dev.off()
    
    jpeg(paste('./Feature_figures/',feat,'_feature_plot.jpg',sep=''))
    plot <- FeaturePlot(combined, features=c(feat), min.cutoff='q9')
    print(plot)
    dev.off()
    
    jpeg(paste('./Feature_figures/',feat,'_split_feature_plot.jpg',sep=''),height=800,width=1500)
    plot <- FeaturePlot(combined, features=c(feat), min.cutoff='q9',split.by='samp')
    print(plot)
    dev.off()
    }
    )
    if(inherits(run, 'try-error')) { next }
   }
  }
   
```

```{r}
save.image('data.RData')
print('Done')
```

